Show downloaded models, improve error handling

cadenmackenzie opened the pull request:
I wanted a way to show which models have already been downloaded. Ultimately, I would like to add the ability to manage those models such as deleting them. Might make more sense in a sidebar where you chose the model and can manage it.

I also added more robust handling of errors in index.html to safely access them as I saw some warnings in the main branch when no errors were set.

In index.js I combined error handling into a single flow for the populateSelector method using response.json() instead of manually parsing. I also added a helper method to set the error to reduce duplicate code. Something I should have done with my last PR.

Open to suggestions on these changes.

cadenmackenzie commented:
I am seeing now some of the robust error handling was intentional and merged in [here](https://github.com/exo-explore/exo/commit/6d12deab2ab158e527e402fb9dfc8d7b66f7cb3d). I can change back if that is preferred.

dtnewman commented:
+1 on the concept, but this is an issue:

![image](https://github.com/user-attachments/assets/dc07e9a9-13f9-42e3-8607-e8e190845b69)

dtnewman commented:
> > +1 on the concept, but this is an issue:
> > ![image](https://private-user-images.githubusercontent.com/3669803/386011807-dc07e9a9-13f9-42e3-8607-e8e190845b69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTAwODEsIm5iZiI6MTczMTU0OTc4MSwicGF0aCI6Ii8zNjY5ODAzLzM4NjAxMTgwNy1kYzA3ZTlhOS0xM2Y5LTQyZTMtODYwNy1lOGUxOTA4NDViNjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTRUMDIwMzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWRiYzZjN2Q1ZTM2ODdjNmYyNzljYmM4Y2NjMjBmNjQ4Zjg4YjkzYmNjMTc2NmI5YjJhMDBlYTgwOTcxY2E0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.XjPC9aH5p_Ny84CxJ3YuuBvs0ebuk8Kvm4OHL2anwRM)
> 
> Oh wait, that's in Safari. It actually works on Chrome and Firefox. Still, should probably fix before merge.
> 
> ![image](https://private-user-images.githubusercontent.com/3669803/386012530-233185ef-93ba-486d-96f9-f14ebc8a7fb8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTExODgsIm5iZiI6MTczMTU1MDg4OCwicGF0aCI6Ii8zNjY5ODAzLzM4NjAxMjUzMC0yMzMxODVlZi05M2JhLTQ4NmQtOTZmOS1mMTRlYmM4YTdmYjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTRUMDIyMTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzU0OWI5MGUyMzE4Mzc2Zjg4NGMwODFmZTgyMDcxZmM4MjdhMjk5MGU0MmRlYzE3ZmUyY2JmMjVmZDBlMTQ0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gfRqB85UJuz0kzqV9UAenqS1DOt7BZaNLJ8xUZNv6gw)

I fixed these issue in https://github.com/cadenmackenzie/exo/pull/1/files

cadenmackenzie commented:
Thanks for finding that! I went ahead and merged it in.

AlexCheema reviewed (changes requested):
What a great idea!
Definitely worthy of a $200 retrospective bounty.

I think the implementation right now is not great -- I'd prefer if we lean on the existing `HFShardDownloader` since that knows how to do downloads. You might need to add a new abstract method to `ShardDownloader` something like `get_shard_download_status` and it should use the same functionality already in `hf_helpers.py` to check what the percentage completion is.

Also, a few suggest changes to the UI: if the model is partially downloaded (percentage > 0) then display the % next to the model name in the UI.

cadenmackenzie commented:
Awesome!

Got it working with the suggested changes. Want to do some testing in the morning then will update this PR.

cadenmackenzie commented:
Added new abstract method in ShardDownloader, implemented get_shard_download_status in HFShardDownloader leaning on get_local_snapshot_dir, get_weight_map, get_allow_patterns helper functions and then checks the percentage of model downloaded for models where local files are found. Removed old function for checking percentage.

Currently shows "downloaded" for fully downloaded models and "X% downloaded" for models not fully downloaded in dropdown.

I don't love how this is being refreshed using the modelPoolInterval because it lags a little for models that are actively being downloaded so might work on how to improve that.

I think it could be worth moving this logic as well as active downloads to a sidebar. I like how active downloads are being shown in the chat when initiated but if we moved that to a sidebar, it could be a centralized place to view all models, choose a model, see activity of downloaded models, and being able to remove local downloads to free up space. Similar to how you can remove local models in LM Studio.

Open to suggestions.

AlexCheema commented on exo/download/hf/hf_shard_download.py:
```diff
@@ -77,3 +87,55 @@ async def wrapped_progress_callback(event: RepoProgressEvent):
+        local_sizes = {}
+        
+        for pattern in patterns:
+            if pattern.endswith('safetensors') or pattern.endswith('mlx'):
```
This doesn't quite work. A pattern can match multiple files.

AlexCheema commented on exo/download/hf/hf_shard_download.py:
```diff
@@ -77,3 +87,55 @@ async def wrapped_progress_callback(event: RepoProgressEvent):
   def on_progress(self) -> AsyncCallbackSystem[str, Tuple[Shard, RepoProgressEvent]]:
     return self._on_progress
+
+  async def get_shard_download_status(self) -> Optional[Dict[str, float]]:
```
Generally there seems to be a lot of duplication between this and other functions. I think a refactor should be done here.

AlexCheema reviewed (changes requested):
Fix the download status code. Otherwise good.

cadenmackenzie commented:
Found an issue in handle_model_support that was creating HFShardDownloader without quick_check=true so it was starting download of models when being checked for download percentage. 

Will work on get_shard_download_status refactor

cadenmackenzie commented:
Hi @AlexCheema can you give me some more insight into what you want to be refactored? I initially thought we could reuse some of the download percentage logic that is happening during download but as far as I can tell, that is only checking against the remote during download in download_file(). Would you want to pull some of that logic out into another helper function to use in get_shard_download_status() or am I missing something?

AlexCheema commented:
> Hi @AlexCheema can you give me some more insight into what you want to be refactored? I initially thought we could reuse some of the download percentage logic that is happening during download but as far as I can tell, that is only checking against the remote during download in download_file(). Would you want to pull some of that logic out into another helper function to use in get_shard_download_status() or am I missing something?

Yeah I think pulling some of that logic out would make sense. I think it's in general in need of a good refactor if you want to take a go at that and will bump up the bounty to $300.

cadenmackenzie commented:
Hi Alex, please review. I didn't modify download_file much but added a check using the new helper method.

Modified the get_shard_download_status to lean on the helper methods to calculate the overall percentage of files and return that. Also updated chatgpt_api to use that overall instead doing percent calculation there. Should fix the pattern matching issue that you identified as well.

cadenmackenzie commented:
@AlexCheema pinging for your review

AlexCheema commented:
Please resolve conflicts and ping me again. @cadenmackenzie

AlexCheema commented:
Did you test this after the merge? Models aren't loading and getting syntax errors.

<img width="715" alt="Screenshot 2024-11-22 at 19 28 48" src="https://github.com/user-attachments/assets/f9bc11fb-552e-464f-9a6e-ab4b7ec5adac">

AlexCheema commented:
Please assign me when fixed and ready for me to review @cadenmackenzie

cadenmackenzie commented:
Hi Alex, my apologies. I did test it and it was working for me even without the inference_engine_classes defined. Not sure why.

I fixed it and installed Pylance.

cadenmackenzie commented:
@AlexCheema merged in side bar changes and delete functionality and added functionality to use server sent events to show each model as its checked. Also shows loading icon in side bar until the first model is loaded. [Quick demo](https://www.loom.com/share/3b1617f6f90248928c38711dbe298e98)

AlexCheema commented:
We need to be very careful about the way periodic requests to `/modelpool` are done. It's a very heavy request so we shouldn't stack them up. Can you make it so only one request happens at a time? One pattern that would work is an while loop with `asyncio.sleep` in between iterations

cadenmackenzie commented on exo/api/chatgpt_api.py:
@AlexCheema did it take a whole minute for all of them to populate? Or a whole minute for the first one to show up?

cadenmackenzie replied:
@AlexCheema I made a change to pull the initial list of models when the page loads and show a loading indicator for each model until it has been checked. They will update as they are checked. Let me know if you have any suggestions.

cadenmackenzie commented on exo/tinychat/index.js:
```diff
@@ -32,12 +34,23 @@ document.addEventListener("alpine:init", () => {
+      this.populateSelector();
+      this.modelPoolInterval = setInterval(() => {
+        this.populateSelector();
+      }, 5000);
```
@AlexCheema fixed in this [commit](https://github.com/exo-explore/exo/pull/456/commits/ded80b0ffdf1f7b9e2da21e577fd9aa461f6d311)

AlexCheema commented:
When I delete a model, the following error appears: `An unknown error occurred`
It does actually delete the model however tinychat shows this error.

cadenmackenzie commented:
Trying to [reproduce](https://www.loom.com/share/99e54d42693e4c04b8a7b7d37310a33c?sid=c8a36b64-336a-4155-b20a-91fffe7d09d9) that error but I can't seem it do it. Can you send me your steps? What line is that error being thrown on?

AlexCheema commented:
Okay, I can't reproduce now either so that's fine.

One more thing: could you add something that indicates whether a download is currently in progress? Sometimes a download could fail and you don't realise it or it could be in progress but got stuck. An indication of whether exo is actively trying to download a shard would be super helpful for debugging. @cadenmackenzie

AlexCheema commented:
> @AlexCheema I added a spinner animation to the model in the sidebar that is actively being downloaded. I can modify that to make it a progress bar or say "Downloading" if that is better. Demo [here](https://www.loom.com/share/7bf02fe4b309472eb478d4e6920c2356?sid=c21d5b86-d6e7-4e36-84e0-ad920cb13224). Is that what you were envisioning?
> 
> Merge conflicts from main branch resolved.

I noticed this is being done from the frontend right now.
Is there any way to use the backend as the source of truth to check if a model download is in progress instead? This will be incredibly useful for debugging e.g. if I'm using the API and I want to check what model downloads are in progress.

cadenmackenzie commented:
@AlexCheema right now the frontend is matching what is being returned by `/v1/download/progress` and `/modelpool` to determine if a model download is in progress on the sidebar. do we want to try to combine some of that logic into one endpoint? Or add a check to `/modelpool` to return true or false if the model is being downloaded. `/v1/download/progress` is the current source of truth for checking if a download is in progress and polls every second when a download is in progress.

AlexCheema commented:
> @AlexCheema right now the frontend is matching what is being returned by `/v1/download/progress` and `/modelpool` to determine if a model download is in progress on the sidebar. do we want to try to combine some of that logic into one endpoint? Or add a check to `/modelpool` to return true or false if the model is being downloaded. `/v1/download/progress` is the current source of truth for checking if a download is in progress and polls every second when a download is in progress.

That works! I just wanted to make sure the source of truth for this was the server, not the frontend.

AlexCheema commented:
> can you clarify what you mean? Do you want the logic to be combined into one endpoint or are you okay that the server is the source of truth using the two endpoints and the matching of results on the frontend as it is right now? @AlexCheema

Unifying into one endpoint would be better.

I also added an endpoint to trigger a download here: https://github.com/exo-explore/exo/pull/526
It would be great if that could be triggered from the UI.

cadenmackenzie commented:
@AlexCheema I got a working version of the endpoints combined but I have some concerns. If the endpoints are combined, the updates for the Download Progress Section rely on a response from /modelpool which is called every 5 seconds and checks the status of _every_ model. Ultimately these feel like processes that should be kept in their own endpoints because /modelpool is checking each local model against the remote whereas /download/progress is using the node.node_download_progress during a download. If anything I think we could use the responses from /download/progress to update the sidebar for the model being downloaded so it is more real-time. Let me know if I am missing something as to why you would want these combined into one endpoint.

cadenmackenzie commented:
Also added the trigger to start a download for partially downloaded or not downloaded models using the download endpoint. Demo [here](https://www.loom.com/share/4191200f8eee45cb8ac4f3cf7c6153c9?sid=3f4bb425-f487-4e79-84c4-4b77766cc70e). @AlexCheema

AlexCheema commented:
> @AlexCheema I got a working version of the endpoints combined but I have some concerns. If the endpoints are combined, the updates for the Download Progress Section rely on a response from /modelpool which is called every 5 seconds and checks the status of _every_ model. Ultimately these feel like processes that should be kept in their own endpoints because /modelpool is checking each local model against the remote whereas /download/progress is using the node.node_download_progress during a download. If anything I think we could use the responses from /download/progress to update the sidebar for the model being downloaded so it is more real-time. Let me know if I am missing something as to why you would want these combined into one endpoint.

Seems good. No need to merge the endpoints.

AlexCheema commented:
Merged. Please email alex@exolabs.net with your ethereum address to claim $300 retrospective bounty
